File: java/src/main/java/com/pqc/mlkem/Field.java
package com.pqc.mlkem;

/**
 * Finite field arithmetic modulo Q = 3329.
 */
public final class Field {

    public static final int Q = 3329;

    private Field() {}

    /**
     * Reduce a mod q into range [0, q).
     */
    public static int mod(int a, int q) {
        int r = a % q;
        return r < 0 ? r + q : r;
    }

    public static int fieldAdd(int a, int b) {
        return mod(a + b, Q);
    }

    public static int fieldSub(int a, int b) {
        return mod(a - b, Q);
    }

    public static int fieldMul(int a, int b) {
        return mod(a * b, Q);
    }

    /**
     * Modular exponentiation: base^exp mod Q.
     */
    public static int fieldPow(int base, int exp) {
        base = mod(base, Q);
        int result = 1;
        for (int i = 0; i < exp; i++) {
            result = fieldMul(result, base);
        }
        return result;
    }
}

File: python/mlkem/field.py
"""Finite field arithmetic modulo Q = 3329."""
from __future__ import annotations

Q = 3329


def mod_q(a: int) -> int:
    """Reduce a modulo Q into range [0, Q)."""
    return a % Q


def field_add(a: int, b: int) -> int:
    """Add two field elements modulo Q."""
    return (a + b) % Q


def field_sub(a: int, b: int) -> int:
    """Subtract two field elements modulo Q."""
    return (a - b) % Q


def field_mul(a: int, b: int) -> int:
    """Multiply two field elements modulo Q."""
    return (a * b) % Q


def field_pow(base: int, exp: int) -> int:
    """Raise a field element to a power modulo Q."""
    return pow(base, exp, Q)
